### ISSUE TITLE
@SuppressWarnings("initialization") stopped working between 3.0.1 and 3.2.0

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#3212]
Was this a deliberate change?

```java
class T {
  @SuppressWarnings("initialization")
  T() {
    f(this);
  }

  void f(T t) {}
}
```

```
$ ./checker-framework-3.0.1/checker/bin/javac -processor Nullness T.java
... OK
$ ./checker-framework-3.2.0/checker/bin/javac -processor Nullness T.java
T.java:4: error: [method.invocation.invalid] call to f(T) not allowed on the given receiver.
    f(this);
     ^
  found   : @UnderInitialization(T.class) @NonNull T
  required: @Initialized @NonNull T
T.java:4: error: [argument.type.incompatible] incompatible types in argument.
    f(this);
      ^
  found   : @UnderInitialization(T.class) @NonNull T
  required: @Initialized @NonNull T
2 errors
```

**Comment by wmdietl** [comment]
It was done in #3115.
I agree that this example should continue to be suppressed.
The problem is that we don't have a way to tell whether those errors came because of the nullness or the initialization hierarchy.
The error key doesn't contain `initialization`, because these are generic error keys, so they are no longer suppressed.

I had a quick look whether there is a simple way to add `initialization` if the error is caused by the initialization hierarchy, but didn't find one.

@mernst It might be best to roll back #3115 until we have a complete solution.

**Comment by mernst** [comment]
The problem is that there is no way to distinguish errors issued by the initialization part of the Nullness Checker from the nullness part of the Nullness Checker.  This is a flaw in the Nullness Checker's implementation.

In 3.0.1, `@SuppressWarnings("initialization")` was equivalent to `@SuppressWarnings("nullness")`.  This suppresses a lot of unintended warnings.  I would prefer not to go back to that state of affairs, if possible.
Liam, is it possible for you to use `@SuppressWarnings("nullness")` instead, or is that too big a change?

**Comment by cushon** [comment]
Thanks for the context!

> Liam, is it possible for you to use `@SuppressWarnings("nullness")` instead, or is that too big a change?

Yes, that's definitely possible. I also have a script to scrape diagnostics and add the corresponding `@SuppressWarnings` for individual diagnostic strings.

So I have a way to unblock upgrading to 3.2.0, I just wanted to understand the change before adding/changing a bunch of suppressions, since this affected a relatively large number of files.

**Comment by wmdietl** [comment]
So on the one hand I agree that we don't want `@SuppressWarnings("initialization")` to suppress unrelated nullness errors.
On the other hand, using `@SuppressWarnings("initialization")` in these cases here better expresses the intent of suppressing initialization errors.
Using `@SuppressWarnings("initialization")` is what we want people to use once the checks are implemented correctly.
So it seems like unnecessary churn to now force them to use `@SuppressWarnings("nullness")` and then again force them to change to `@SuppressWarnings("initialization")`.

**Comment by mernst** [comment]
I agree that `@SuppressWarnings("initialization")` is more descriptive.  It would be better if the Checker Framework supported that.

But, do we have any concrete plan for supporting that?  I think I wouldn't want to just leave `@SuppressWarnings("initialization")` suppressing everything for an indefinite period.

**Comment by cushon** [comment]
I went ahead and mechanically replaced all of our `initialization` suppressions with `nullness` ones to unblock migrating to 3.2.0.

Having a more precise way to suppress initialization errors would be nice, but in practice I'm not sure I'd unwind that cleanup. It's fine with me if you want to close this bug.

**Comment by mernst** [comment]
Thanks for following up.
I agree entirely that we want a more precise way to suppress initialization errors.
I'll close this issue, though.

**Comment by wmdietl** [comment]
If we ever get to #481 we should have more information what sub-hierarchy violated a property and can then include that information for suppression.
